Refactor Compose

pollyannagogolook opened the pull request:
1. Revise naming issue in data layer according to CRUD.
2. Use constructor modifier to control layout of composable function.
3. Revise naming issue in compose.
4. Use mutableState in viewModel rather than stateflow.
5. Use callback functions and remember to restrict recomposition scope in composable functions.

pollyannagogolook committed:
Feat: Draw a search bar with compose

- add dependencies for compose in gradle
- Remove viewBinding in MainActivity
- create a composable function searchbar without searching history in HomeScreen

pollyannagogolook committed:
Feat: Inject PagingData in composable function

- Use collectAsLazyPagingItem for submitting data to PagingAdapter
- Layout of each card is not finished yet.
- Layout of searching history is not finished yet.

pollyannagogolook committed:
Feat: Enable search image in search bar with compose

- inject viewModel in comosable function with hilt.
- call viewModel's function in search bar composable function.
- collect stateflow from viewModel as state in composable function.
- present searching history as lazy column under search bar.

pollyannagogolook committed:
Feat: Draw linear layout of image with Compose

- Create Type.kt to standardize text template
- Create a LinearImageScreen to present linear layout arrangement.
- In LinearImageScreen, I use constraint layout.
- Create getNumberText function to perform as BindingAdapter function numberText.
- In preview function, use fake data fakeHit.

pollyannagogolook committed:
Feat: Change layout management in one composable function SingleImageScreen

- Add isLinear as a parameter to change layout management by injectting state.
- When layout is grid, the card will only show comment and likes.
- Call SingleImageScreen in HomeScreen

pollyannagogolook committed:
Feat: Add init function in MainViewModel

- set getDefaultLayout function and loadAllImage function in init function at MainViewModel
- when getting default layout from remote, let isLinear collect the value then pass to compose as state

- Add null-safety check in SingleImageScreen

- There is still a bug: when click on search bar, app will crash(Seems to be related to Glide)

pollyannagogolook committed:
Fix: Crash related to Glide when user scrolls down

- Glide compose dependency seems unstable, replace it with coil dependency at gradle.
- Replace GlideImage with Image whose painter uses coil to load image in SingleImageScreen

pollyannagogolook committed:
Fix: Layout arrangement is not correct.

- Replace constraint layout with row and column.
- Leaving padding horizontally on numberIcon composable function.

pollyannagogolook committed:
Feat: Add LoadingScreen in HomeScreen

- Replace collect with assigning image stateflow in ViewModel. Because when user search images, they will not expect to see original images.
- Create a LoadingScreen to present loading animation.

pollyannagogolook committed:
Feat: When click on toggle btn, should change to grid layout.

- Add Switch component and linear layout drawable as toggle btn to linear layout, and default setting is turning on.
- Divide SingleImageScreen to 3 parts: UserInfoComposable, MainImageComposable and ActionRowComposable to increase readability.
- When isLinear is false, use LayVerticalGrid in HomeScreen, and each card's width is 0.5f.

pollyannagogolook committed:
Feat: When open the app, will show loading page

- To let user know the app is fetching data when they enter, show loading page while execueting init function,

pollyannagogolook committed:
Fix: Resolve git conflict when merge feature_compose to develop branch

- Merge branch 'develop' into feature_compose
- solve git conflict in this branch
- remove redundant code and useless variables

pollyannagogolook committed:
fix: change naming from screen to view

pollyannagogolook committed:
Fix: Only pass viewModel to HomeScreen as parameter to avoid memory leak.

- remove viewModel as constructor in Search, LoadingView, SingeImageView composable functions.
- change variable's type from stateflow to state in viewModel.
- create LazyPagingContent composable function to restrict recomposition scope.
- use remember to restrict recomposition scope in HomeScreen.

pollyannagogolook committed:
Fix: use constructor modifier to control composable function's layout

pollyannagogolook committed:
Fix: search history didn't record user's searching.

- set callback function saveSearchHistory in SearchBar to trigger viewModel's function in HomeScreen.
- set showSearchSuggestion callback function in SearchBar to trigger viewModel' function in HomeScreen.
- Revise naming in data layer according to CRUD rule.

pollyannagogolook committed:
Fix: Divide 2 composable function when isLinear is false in SingleImageView

- Create LinearActionRow and GridActionRow composable functions.
- Use constructor to control composable function's layout
- Create LimitedLineText composable function to limited text line to avoid interrupt layout management.

christclin commented on app/src/main/java/com/pollyannawu/gogolook/MainViewModel.kt:
```diff
@@ -29,31 +21,37 @@ class MainViewModel @Inject constructor(private val repository: Repository) : Vi
-        get() = _images.filterNotNull()
 
+    var images = MutableStateFlow<PagingData<Hit>>(PagingData.empty())
+    private val _images = MutableStateFlow<PagingData<Hit>>(PagingData.empty())
```
這裡的 _images 還是要繼續用，而 images 應該是 StateFlow (不是 MutableStateFlow)，以面從外部非預期改動

christclin commented on app/src/main/java/com/pollyannawu/gogolook/MainViewModel.kt:
```diff
@@ -29,31 +21,37 @@ class MainViewModel @Inject constructor(private val repository: Repository) : Vi
-    private val _isLinear = MutableStateFlow<Boolean>(true)
-    val isLinear: StateFlow<Boolean> = _isLinear.asStateFlow()
+    var isLinear = mutableStateOf<Boolean>(true)
+    private set
```
這一系列的 mutableState 用法，有幾個建議
1. 使用 val，因為 state 本身不需要重新 assign，而是使用 state.value 更新
2. 建議還是要有 private 和 public 一組，以面非預期從外部更新
```
private val _isLinear = mutableStateOf(true)
val isLinear: State<Boolean> = _isLinear
```

christclin commented on app/src/main/java/com/pollyannawu/gogolook/compose/home/HomeScreen.kt:
```diff
@@ -0,0 +1,283 @@
+    }
+
+    // use in home pager view
+    val lazyPagerItems = lazyPagingContent(pagingFlow = viewModel.images)
```
這裡不需要多一層 lazyPagingContent，直接這樣用就行
```
val images = viewModel.images.collectAsLazyPagingItems()
```

pollyannagogolook replied:
這邊已經做修正。另外想跟大大確認一下，這邊不用另外使用 lambda 的方式限縮 recompose scope 的原因是因為這個 `collectAsLazyPagingItems()` 裡面已經使用 LaunchEffect 讓我們不太需要擔心 recomposition 的問題是嗎？
昨天在這塊搞了蠻久，不知道在 paging 的狀態下，到底該怎麼限制 recomposition scope ，偏偏這又是最上層。

christclin commented on app/src/main/java/com/pollyannawu/gogolook/compose/home/HomeScreen.kt:
```diff
@@ -0,0 +1,283 @@
+@Composable
+fun LayoutToggleButton(
+    modifier: Modifier = Modifier,
+    isLinear: Boolean,
```
這裡使用 callback，可以讓 recompose 只發生在這一層
```
isLinear: () -> Boolean
```

pollyannagogolook replied:
這邊有做了一點修改，已經限制 isLinear 的作用域了。不過這邊之所以會這樣設計，是因為 isLinear 除了影響 toggleButton 本身，也會影響 PagerScreen 的 layout, 以及 singleImageView 的 action row 。我原本是想說如果邏輯本來就是 isLinear 會需要控制這麼多地方，那從 homeScreen 傳值是蠻正常的，反過來說如果都用 toggleBtn callback 回來，我覺得好像效果跟原本差不多。
這邊想問問大大意見～

christclin replied:
你的思路是正確的，如果從最上層開始本來就會 recompose，那好像的確不需要，不過
1. HomeScreen 還是有些地方不需要因為 isLinear 變動而 recompose，利如 SearchBar
2. 我們考慮到你的 Composable 未來的重用性，無論如何在每個地方建立好習慣是會比較好的

christclin commented on app/src/main/java/com/pollyannawu/gogolook/compose/home/SingleImageView.kt:
```diff
@@ -0,0 +1,284 @@
+}
+
+@Composable
+fun MainImageView(painter: Painter, modifier: Modifier = Modifier) {
```
modifier 建議放在第一個參數

jason-gogolook commented on app/src/main/java/com/pollyannawu/gogolook/compose/ImageSearcherApp.kt:
```diff
@@ -0,0 +1,37 @@
+    navController: NavHostController,
+    viewModel: MainViewModel
+){
+   val activity = (LocalContext.current) as Activity
```
這個 activity 沒有被使用到

jason-gogolook commented on app/src/main/java/com/pollyannawu/gogolook/compose/Screen.kt:
```diff
@@ -0,0 +1,12 @@
+
+sealed class Screen(
+    val route: String,
+    val navArgument: List<NamedNavArgument> = emptyList()
```
`navArgument` 這個參數目前沒有被使用，是為了之後設計預留嗎？

pollyannagogolook replied:
沒錯的，雖然我現在專案弄的很簡單，所以嚴格來說根本連 navigation 都不太需要，但是想說我就趁機練習一下 navigation 在 compose 的應用這樣。

jason-gogolook commented on app/src/main/java/com/pollyannawu/gogolook/compose/Screen.kt:
```diff
@@ -0,0 +1,12 @@
+import androidx.navigation.NavType
+import androidx.navigation.navArgument
+
+sealed class Screen(
```
好奇問會想使用 `seal class` 的目的性是什麼？

pollyannagogolook replied:
目前因為專案上只有一個 screen ，所以 sealed class 確實沒有太大意義。
不過想說如果 compose 要比較好維護的話，看 sample code 還有 stackoverflow 的說明（https://stackoverflow.com/questions/69686087/why-use-sealed-class-and-make-object-in-navigation-kotlin-jetpack-compose）是把所有頁面還有各自的 route 做成一個 sealed class 比較好管理，尤其是未來如果有頁面需要帶 nav argument 的情況。

jason-gogolook commented on app/src/main/java/com/pollyannawu/gogolook/data/model/search_history/SearchHistoryDataSource.kt:
```diff
@@ -27,7 +25,7 @@ class SearchHistoryDataSource @Inject constructor(private val app: Application)
 
 
-    fun saveSearchQuery(query: String) {
+    fun insetSearchHistory(query: String) {
```
insertSearchHistory ?

pollyannagogolook committed:
Refactor: use private variable in viewModel to avoid value is changed outside.

- set mutableState variable as  private type
- use State varialbe to interact with view

pollyannagogolook committed:
Refactor: restrict recompose scope in HomeScreen

- use callback to restrict recompostion in toggleLayout.
- move modifier as the first parameter in composable function.
- typo: inset to insert
- remove redundant variable in ImageSearchApp.

pollyannagogolook committed:
Refactor: use callback function in composable function to restrict composable function
